compiler/rustc_session/src/session.rs

```diff
@@ -268,6 +268,14 @@ impl From<&'static lint::Lint> for DiagnosticMessageId {
     }
 }
 
+impl std::ops::Deref for Session {
+    type Target = rustc_errors::Handler;
+
+    fn deref(&self) -> &Self::Target {
+        self.diagnostic()
+    }
+}
```

Comment:
This is not very clean IMO and prevents deref to any other type in the future.

Reply:
Yea... this always was the "obvious" way to eliminate the duplication and we didn't do it on purpose.

we could add a `diag()` method on `TyCtxt` and `Session` and "just" use it everywhere. This would change all `tcx.sess.span...` to `tcx.diag().span...`, so it isn't really more verbose.

Reply:
Hmm, I wonder if that should go through MCP - it would be a lot of work and I'd hate to throw it away because someone didn't like the new API.

Reply:
Yea, do an MCP or bring it up with @estebank directly before opening an MCP